### PR TITLE
Refactored OOA

### PR DESCRIPTION
**Pull request by powertashton**
student_manage_edit
student_manage_delete
student_manage_add
staff_manage_addProcess
staff_manage_edit
staff_manage_delete
staff_manage_add
cas_supervisor
cas_supervisor_invite
cas_student_reflections_delete
cas_student_reflections_add
cas_student_myCommitments_edit
cas_student_myCommitments_delete
cas_student_myCommitments_add
cas_student_interview3
cas_student_interview2
cas_adviseStudents_details

manifest

These files have been altered to OO PHP from Legacy. 
I have taken the old legacy forms, updated so that they use the new form object. Changes on the pages above.

### CODE REVIEW

File: IB Diploma/CHANGELOG.txt [hunk 1]
```diff
@@ -1,5 +1,9 @@
 CHANGELOG
 =========
+v1.0.11
```
Comment (SKuipers) [inline IB Diploma/CHANGELOG.txt]: Feel free to kick this up to the [next minor version](https://semver.org/), v1.1.00 👍 I think the refactoring you've done is big enough in scope.

File: IB Diploma/cas_adviseStudents_details.php [hunk 2]
```diff
@@ -159,50 +164,50 @@
...
                         $count = 0;
-                        $rowNum = 'odd';
+                        $valuesNum = 'odd';
                         $intended = array();
                         $complete = array();
-                        while ($row = $result->fetch()) {
+                        while ($values = $result->fetch()) {
                             if ($count % 2 == 0) {
-                                $rowNum = 'even';
+                                $valuesNum = 'even';
```
Comment (SKuipers) [inline IB Diploma/cas_adviseStudents_details.php]: There's a `colorOddEven` css class that can replace any of the `$count % 2` etc code for these kinds of tables. Even better, there's a whole new set of Table OO classes. Perhaps that could be step 2 for your refactoring? 😃 give a shout if you're interested and I'd be happy to share the docs & some pointers.

File: IB Diploma/cas_adviseStudents_details.php [hunk 3]
```diff
@@ -373,38 +378,24 @@
...
-									<input type="hidden" name="gibbonPersonID" value="<?php echo $gibbonPersonID ?>">
-									<input type="hidden" name="address" value="<?php echo $_SESSION[$guid]['address'] ?>">
-									<input type="submit" value="Submit">
-								</td>
-							</tr>
-						</table>
-					</form>
-					<?php
+
+					$form = Form::create('action', $_SESSION[$guid]['absoluteURL'].'/modules/'.$_SESSION[$guid]['module'].'/cas_adviseStudents_detailsStatusProcess.php', "post");
```
Comment (SKuipers) [inline IB Diploma/cas_adviseStudents_details.php]: The 'action' in `Form::create('action', ...` is the ID of the table, it helps to add something descriptive like 'casStatus'. It not only helps readability-wise, but also to keep any of the JS separate when there could be more than one form or table on the page.

File: IB Diploma/cas_adviseStudents_details.php [hunk 4]
```diff
@@ -680,137 +561,65 @@
...
-                                if ($rowInterview["2_outcome$i"] != '') {
-                                    $commitments = explode(',', $rowInterview["2_outcome$i"]);
-                                    foreach ($commitments as $commitment) {
-                                        $where .= 'ibDiplomaCASCommitmentID='.$commitment.' OR ';
-                                    }
-                                    $where = substr($where, 0, -4);
+								
+								$prepopulate = '';
+                                if ($valuesInterview["2_outcome".$i] != '') {
+                                	echo "2_outcome".$i;
```
Comment (SKuipers) [inline IB Diploma/cas_adviseStudents_details.php]: Is this a debug output? It seems a bit out of place, hard to tell in the diff-view though 🤔

File: IB Diploma/cas_adviseStudents_details.php [hunk 5]
```diff
@@ -680,137 +561,65 @@
...
+									$row->addLabel('casStatusSchool', __('CAS Status'));
+									$row->addSelect('casStatusSchool')->fromArray(array('At Risk' =>__('At Risk'), 'On Task' => __('On Task'), 'Excellence' =>__('Excellence')))->selected($casStatusSchool)->isRequired();
+						
+                            
+							$row = $form->addRow();
+								$row->addFooter();
+								$row->addSubmit();
+							echo $form->getOutput();
+							
+							// HACK: Otherwise FastFinder width overrides this one :(
```
Comment (SKuipers) [inline IB Diploma/cas_adviseStudents_details.php]: Love that you transplanted the comment and `:(` along with the fix 😆 We'll have to make a note to tackle this one as we start in on the responsive refactoring 😄

File: IB Diploma/cas_student_myCommitments_delete.php [hunk 6]
```diff
@@ -66,8 +68,15 @@
...
-                $row = $result->fetch();
+                $values = $result->fetch();
+                 	echo "<div class='linkTop'>";
+                 	echo "<a href='".$_SESSION[$guid]['absoluteURL'].'/index.php?q=/modules/'.getModuleName($_GET['q'])."/cas_student_myCommitments.php'>".__($guid, 'Back').'</a>';
+                 	echo '</div>';
+
+                $form = DeleteForm::createForm($_SESSION[$guid]['absoluteURL'].'/modules/'.$_SESSION[$guid]['module'].'/cas_student_myCommitments_deleteProcess.php?ibDiplomaCASCommitmentID=$ibDiplomaCASCommitmentID'."&ibDiplomaCASCommitmentID=".$_GET['ibDiplomaCASCommitmentID']);
+            	echo $form->getOutput();
                 ?>
+<!--
```
Comment (SKuipers) [inline IB Diploma/cas_student_myCommitments_delete.php]: Looks like some dead code from the old form, now that you're done the refactor it's a great time to kill it 😃

File: IB Diploma/cas_supervisor_invite.php [hunk 7]
```diff
@@ -57,6 +61,9 @@
             echo '</h3>';
 
             ?>
+
+
+<!--
```
Comment (SKuipers) [inline IB Diploma/cas_supervisor_invite.php]: Some more dead code :)

File: IB Diploma/staff_manage_addProcess.php [hunk 8]
```diff
@@ -41,14 +41,17 @@
...
 } else {
     //Proceed!
     $gibbonPersonID = $_POST['gibbonPersonID'];
     $role = $_POST['role'];
 
     if ($gibbonPersonID == '' or $role == '') {
+    	
         //Fail 3
         $URL = $URL.'&return=error3';
+        var_dump ($gibbonPersonID);
```
Comment (SKuipers) [inline IB Diploma/staff_manage_addProcess.php]: 👀  Debug output?

File: IB Diploma/student_manage_add.php [hunk 9]
```diff
@@ -35,117 +38,37 @@
...
+	
+	$row = $form->addRow();
+		$row->addLabel('gibbonPersonID', __('Students'));
+		$row->addSelectStudent('gibbonPersonID',$_SESSION[$guid]['gibbonSchoolYearID'], array('allStudents' => true, 'byName' => false, 'byRoll' => true, 'showRoll' => true))->selectMultiple()->isRequired();
+		
+	$data = array('gibbonSchoolYearID' => $_SESSION[$guid]['gibbonSchoolYearID']);
+	$sql = "SELECT gibbonSchoolYearID as value,name FROM gibbonSchoolYear ORDER BY sequenceNumber";
+	$row = $form->addRow();
+		$row->addLabel('gibbonSchoolYearIDStart', __('Start Year'));
+		$row->addSelect('gibbonSchoolYearIDStart')->fromQuery($pdo, $sql)->placeholder(__('Please select...'))->isRequired();
```
Comment (SKuipers) [inline IB Diploma/student_manage_add.php]: Luckily the Select class will auto-add the `__('Please select...')` placeholders if the field is required. Helps cut down on a bit of code, and also helps if the text ever changes in the future.

**Review by SKuipers — CHANGES_REQUESTED** [review]
Hi Ashton,

This PR is looking great! You certainly have a handle on how the Form classes work 👍 I've added a few notes and suggestions. Is this a one-off project or would you be interested in taking the next steps for refactoring this module? There's a set of Table OO classes, as well as some Gateway classes that would be a great next step.

A couple other overall notes on code style:
- You'll see a lot of the old code uses tabs for indentation. We've since set a standard at 4 spaces instead of tabs. If your code editor has an automatic convert tabs to spaces tool, it helps to run this on any files you change to improve things as we go.
- If you see any `__()` methods in the code still using `__($guid, ` feel free to remove the $guid parameter, it's deprecated now and we're aiming to remove it from the codebase as a whole.

File: IB Diploma/staff_manage_addProcess.php (hunk 8 above)
Comment (powertashton) [reply IB Diploma/staff_manage_addProcess.php]: Yep, sorry.

File: IB Diploma/CHANGELOG.txt (hunk 1 above)
Comment (powertashton) [reply IB Diploma/CHANGELOG.txt]: Done. Thank You 😄

File: IB Diploma/cas_adviseStudents_details.php (hunk 2 above)
Comment (powertashton) [reply IB Diploma/cas_adviseStudents_details.php]: Yes please, that sounds interesting, can you send me the details?

File: IB Diploma/cas_adviseStudents_details.php (hunk 4 above)
Comment (powertashton) [reply IB Diploma/cas_adviseStudents_details.php]: Was indeed a debug output, must've missed it when I was going through first time through, woops, sorry.

File: IB Diploma/cas_adviseStudents_details.php (hunk 5 above)
Comment (powertashton) [reply IB Diploma/cas_adviseStudents_details.php]: Yes, it had me stumped for a while, the `:(` did bring a smile to my face.

**Comment by powertashton** [comment]
Yes I would be interested in taking the next steps for refactoring this module. Thanks

**Review by SKuipers — CHANGES_REQUESTED** [review]
Hi again! This is looking great, sorry for the wait (busy times at the moment!). I've installed it and tested it out, I think it's looking pretty close to merge-ready 👍  I noticed one small thing so-far, the `student_manage_edit.php` page is missing a `$form->loadAllValuesFrom($values);` at the bottom, to pre-fill the existing values. Aside from that, looks good, thanks!

**Comment by SKuipers** [comment]
Oh, also be sure to catch the `$moduleVersion` in `version.php`. The version number tends to exist in four places: manifest.php, version.php, CHANGELOG.txt and CHANGEDB.php

**Review by SKuipers — APPROVED** [review]

**Comment by rossdotparker** [comment]
Ash, it's amazing to see how much progress you've made on this. I've merged it, and now going to do some testing before releasing it on the website. Thanks : )

**Comment by rossdotparker** [comment]
It's looking good. The only issue I can see so far is that adding a commitment with the following inputs (and any other inputs as far as I can see) returns a database error:

Screenshot 2019-03-13 at 9 37 17 PM

Do you want to look for and PR a fix for this, and then I can do some further testing. Thanks : )

**Comment by powertashton** [comment]
Hi, @rossdotparker sorry I can't reproduce that error, on version 17 or 18. Is there something different on your system? Could you send me more debug information so I can double check what's going on?
image
image

**Comment by SKuipers** [comment]
Hi Ashton,

I did a quick test on my system, looks like the error might be [SQL mode related](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html). Here is the error message: `General error: 1364 Field 'goals' doesn't have a default value` which results in an `error2` return value from line 101 of `cas_student_myCommitments_addProcess.php`. Looks like goals is defined as a `NOT NULL` field, so it needs a value passed in, even if that value is an empty string.

Lots of development installs like MAMP will not have the strict setting on, so the error often only turns up in production or on other people's systems 😅

**Comment by SKuipers** [comment]
Here's some ideas for the next steps in refactoring:

- **Breadcrumbs**: check out the [issue here](https://github.com/GibbonEdu/core/issues/696) which has some instructions. We can add the IB module to the list 😃 

- **Code Cleanup**: The `@session_start();` & duplicate DB connections can be deleted, these are now handled by the core. Anytime you see the following at the top of a file it can safely be removed (& PRd)

```
//New PDO DB connection
try {
    $connection2 = new PDO("mysql:host=$databaseServer;dbname=$databaseName", $databaseUsername, $databasePassword);
    $connection2->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    $connection2->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
} catch (PDOException $e) {
    echo $e->getMessage();
}

@session_start();
```

Also these timezone functions can be removed:
```
//Set timezone from session variable
date_default_timezone_set($_SESSION[$guid]['timezone']);
```

- **Refactoring Tables**: Much like the `Form` class, there's a set of `DataTable` classes for handling table data & lists of records. There's [an ooification guide here](https://gist.github.com/SKuipers/e176454a2feb555126c2147865bd0626). These can certainly get trickier than forms, feel free to give a shout on slack with any questions.

**Comment by rossdotparker** [comment]
Hi @powertashton sorry for the delay in replying, but it looks like @SKuipers bailed me out. Let us know how you get on with this : )

**Pull request merged**